src/parser/mod.rs

```diff
@@ -2266,6 +2312,109 @@ impl<'a> Parser<'a> {
         })
     }
 
+    /// Parse an internal data type of clickhouse nullable [1]
+    /// Internal data type of nullable can’t be any of the composite data types Array, Map and Tuple
+    /// Syntax
+    /// ```sql
+    /// Nullable(data_type)
+    /// ```
+    ///
+    /// [1]: https://clickhouse.com/docs/en/sql-reference/data-types/nullable
+    fn parse_click_house_nullable(&mut self) -> Result<DataType, ParserError> {
+        self.expect_keyword(Keyword::NULLABLE)?;
+
+        self.expect_token(&Token::LParen)?;
+        let (inside_type, _) = self.parse_data_type_helper()?;
+        if matches!(
+            inside_type,
+            DataType::Array(_) | DataType::Map(_, _) | DataType::Tuple(_)
+        ) {
+            return parser_err!("Unexpected internal data type of nullable:", inside_type);
+        }
+        self.expect_token(&Token::RParen)?;
+
+        Ok(inside_type)
+    }
+
+    /// Parse clickhouse low cardinality [1]
+    /// Internal data type of low cardinality can be String, FixedString, Date, DateTime and numbers excepting Decimal
+    /// Syntax
+    /// ```sql
+    /// LowCardinality(data_type)
+    /// ```
+    ///
+    /// [1]: https://clickhouse.com/docs/en/sql-reference/data-types/lowcardinality
+    fn parse_click_house_low_cardinality(&mut self) -> Result<DataType, ParserError> {
+        self.expect_keyword(Keyword::LOWCARDINALITY)?;
+
+        self.expect_token(&Token::LParen)?;
+        let (inside_type, _) = self.parse_data_type_helper()?;
+        if !matches!(
+            inside_type,
+            DataType::String(_) | DataType::Date | DataType::Datetime(_)
+        ) {
+            return parser_err!(
+                "Unexpected internal data type of low cardinality:",
+                inside_type
+            );
+        }
+        self.expect_token(&Token::RParen)?;
+
+        Ok(inside_type)
+    }
+
+    /// Parse clickhouse map [1]
+    /// Syntax
+    /// ```sql
+    /// Map(key_data_type, value_data_type)
+    /// ```
+    ///
+    /// [1]: https://clickhouse.com/docs/en/sql-reference/data-types/map
+    fn parse_click_house_map(&mut self) -> Result<(DataType, DataType), ParserError> {
+        self.expect_keyword(Keyword::MAP)?;
+        self.expect_token(&Token::LParen)?;
+        let (key_data_type, _) = self.parse_data_type_helper()?;
+        self.expect_token(&Token::Comma)?;
+        let (value_data_type, _) = self.parse_data_type_helper()?;
+        self.expect_token(&Token::RParen)?;
+
+        Ok((key_data_type, value_data_type))
+    }
+
+    /// Parse clickhouse tuple [1] and nested [2]
+    /// Syntax
+    /// ```sql
+    /// Nested(field_name field_type, ...)
+    /// Tuple([field_name] field_type, ...)
+    /// ```
+    ///
+    /// [1]: https://clickhouse.com/docs/en/sql-reference/data-types/tuple
+    /// [2]: https://clickhouse.com/docs/en/sql-reference/data-types/nested-data-structures/nested
+    fn parse_click_house_data_type_with_subtype<F>(
```

Comment:
I don't think this function needs trailing bracket in its signature? (might be a result of our calling `parse_data_type_helper` upstream instead of `parse_data_type`
 if its not needed, this functionality might be fulfilled already by some variant of `parse_comma_separated`

Reply:
I removed this function aligning all previous changes.